pkg/querier/querier.go

```diff
@@ -527,3 +532,46 @@ func (q *Querier) validateQueryTimeRange(userID string, from *time.Time, through
 
 	return nil
 }
+
+func (q *Querier) checkTailRequestLimit(ctx context.Context) error {
+	userID, err := user.ExtractOrgID(ctx)
+	if err != nil {
+		return err
+	}
+
+	replicationSet, err := q.ring.GetAll()
+	if err != nil {
+		return err
+	}
+
+	// we want to check count of active tailers with only one of the active ingesters since
+	// all of them would be having same number of active tail connections
+	// Note: In worst-case scenario the ingester that we picked would have joined recently which might not have all the tail requests
+	ingesters := make([]ring.IngesterDesc, 0, 1)
```

Comment:
I don't know about this. I could technically have tailing on just some ingesters, because my matchers fits only those ingester, I would prefer something more robust.

How about when opening tailing you add a unique id from the querier, ask all ingesters and count by unique id ?

Reply:
I don't think it works this way? i don't think the matcher is used to determine which ingester to tail, i think all of them are always tailed

Reply:
it use matcher to find matching stream, if there is no matching streams then it won't have any tailer.

Reply:
the tailer is still created regardless and a separate count is kept, the tailer is kept even if it doesn't match any streams as it might match a new stream in the future

Reply:
```func (i *instance) addNewTailer(t *tailer) {
	i.streamsMtx.RLock()
	for _, stream := range i.streams {
		if stream.matchesTailer(t) {
			stream.addTailer(t)
		}
	}
	i.streamsMtx.RUnlock()

	i.tailerMtx.Lock()
	defer i.tailerMtx.Unlock()
	i.tailers[t.getID()] = t
}````

Also I think we want to limit the amount of tail requests not the amount of tailers across streams ?

Reply:
Ho I think I understand `i.tailers` always gets added.